src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/OptimizedCompletionList.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Newtonsoft.Json;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    [JsonConverter(typeof(OptimizedCompletionListJsonConverter))]
+    internal class OptimizedCompletionList : CompletionList
+    {
+        public OptimizedCompletionList(CompletionList innerList) : base(innerList.Items, innerList.IsIncomplete)
+        {
+        }
+
+        public class OptimizedCompletionListJsonConverter : JsonConverter
+        {
+            public static readonly OptimizedCompletionListJsonConverter Instance = new OptimizedCompletionListJsonConverter();
+            private static readonly string TagHelperIconRawJson;
+            private static readonly JsonSerializer DefaultSerializer;
+
+            static OptimizedCompletionListJsonConverter()
+            {
+                DefaultSerializer = JsonSerializer.CreateDefault();
+                TagHelperIconRawJson = JsonConvert.SerializeObject(VSLspCompletionItemIcons.TagHelper);
+            }
+
+            public override bool CanConvert(Type objectType)
+            {
+                return typeof(OptimizedCompletionList).IsAssignableFrom(objectType);
+            }
+            public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+            {
+                var result = DefaultSerializer.Deserialize(reader, objectType);
+                return result;
+            }
+
+            public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer)
+            {
+                var completionList = (CompletionList)value;
+
+                writer.WriteStartObject();
+
+                writer.WritePropertyName("isIncomplete");
```

Comment:
Isn't there a default for this?

Reply:
I assumed `false` would be the default; however, I choose to provide it because the spec doesn't mark it as optional:
![image](https://user-images.githubusercontent.com/2008729/103711292-f676ae00-4f6b-11eb-89fd-3017b0cd36f2.png)

Reply:
Doesn's this indicate the default is false?

Response:

result: CompletionItem[] | CompletionList | null. If a CompletionItem[] is provided it is interpreted to be complete. So it is the same as { isIncomplete: false, items }

Reply:
Technically yes but isn't that dependent on how the corresponding platform deserializes the response? I guess technically we could further optimize and turn the CompletionList into a `CompletionItem[]` if isIncomplete was false.

Reply:
I dunno, the spec says if CompletionItem[] is provided and isIncomplete isn't specified (the normal case), then it should default to false. Doesn't seem like we should have to send it.

Reply:
But one big difference is that one is an array of completion items and the other is an object with an `items` property and an expected `isIncomplete` property at least according to the spec. I bet this would "just work" 90% of the time but I'm hesitant to do it since it's not strictly spec compliant.

Reply:
Talked offline. For now going to leave it as is to not overoptimize this piece but we can definitely revisit later.